spec/datadog/profiling/stack_recorder_spec.rb

```diff
@@ -555,25 +555,39 @@ def has_seen_id_flag(obj)
           end
 
           def create_obj_in_recycled_slot(should_sample_original: false)
-            obj = Object.new
-            sample_allocation(obj) if should_sample_original
-            @recycled_sample_allocation_line = __LINE__ - 1
-
-            # Get the id of the object we're about to recycle
-            obj_id = obj.object_id
-
-            # Force recycle the given object
-            described_class::Testing._native_gc_force_recycle(obj)
+            # Force-recycle 1000 objects.
+            # NOTE: In theory, a single force recycle would suffice but the more recycled slots
+            #       there are to use the more probable it is for a new allocation to use it.
+            recycled_obj_ids = []
+            1000.times do
+              obj = Object.new
+              sample_allocation(obj) if should_sample_original
+              @recycled_sample_allocation_line = __LINE__ - 1
+
+              # Get the id of the object we're about to recycle
+              recycled_obj_ids << obj.object_id
+
+              # Force recycle the given object
+              described_class::Testing._native_gc_force_recycle(obj)
+            end
 
-            # Repeatedly allocate objects until we find one that resolves to the same id
+            # Repeatedly allocate objects until we find one that resolves to the id of one of
+            # the force recycled objects
+            objs = []
             loop do
               # Instead of doing this one at a time which would be slow given id2ref will
```

Comment:
Should we perhaps move this inside the loop? I'm thinking in cases where we get "unlucky", this may cause memory to spike a lot (and may even lead to an OOM kill).

Reply:
It being outside the loop was intentional: I don't want GC to be able to reclaim slots from a previous iteration otherwise we risk entering an infinite loop (e.g. a GC runs between each iteration and during an iteration all 1000 objects get allocated on slots cleaned up by the previous GC).

Given this method forcibly creates 1000 recycled slots and on each iteration we will create 1000 new objects, I'd hope that before deciding to expand heap pages and potentially triggering an OOM it'd decide to simply re-use the free slots it already has.

Reply:
That's fair. What do you think of perhaps limiting the number of iterations? E.g. going from something like

```ruby
loop do
  ...
end
```

to

```ruby
100.times do 
  ...
end
raise 'Could not cause object to be recycled!'
```

I think if by the time we get to some large number of objects, if we haven't managed to trigger the issue, it seems cleaner to fail the test, rather than letting the process balloon and die.

(Only a suggestion, happy to keep the `loop` if you're not convinced!)

Reply:
Really minor: I think it may be worth blowing up explicitly, as right now if it the looping fails, it'll return `nil`, which will still make the specs fail but in a rather non-obvious way.

```
  1) Datadog::Profiling::StackRecorder#serialize heap samples and sizes when enabled on Rubies supporting rb_gc_force_recycle enforces seen id flag on untracked objects that replace tracked recycled objects
     Failure/Error: expect(has_seen_id_flag(recycled_obj)).to be true
     
       expected true
            got false
```

Reply:
Makes sense! Added an explicit failure:

```
  5) Datadog::Profiling::StackRecorder#serialize heap samples and sizes when enabled on Rubies supporting rb_gc_force_recycle correctly handles lifecycle of objects on recycled slots that get sampled
     Failure/Error: raise 'could not allocate an object in a recycled slot'
     
     RuntimeError:
       could not allocate an object in a recycled slot
```